lib/Conversion/CGGIToJaxite/CGGIToJaxite.cpp

```diff
@@ -0,0 +1,306 @@
+    return op->emitOpError() << "Found less than 2 ops in function input. Did "
+                                "the AddServerKeyArg pattern fail to run?";
+  }
+  auto block_args_list = op->getParentOfType<func::FuncOp>()
+                             .getBody()
+                             .getBlocks()
+                             .front()
+                             .getArguments();
+  auto it = block_args_list.begin();
+  while (!mlir::isa<jaxite::ParamsType>(it->getType())) {
+    it++;
+  }
+  if (it == block_args_list.end() ||
+      !mlir::isa<jaxite::ParamsType>(it->getType())) {
+    return op->emitOpError() << "Cannot find Params argument. Did the "
+                                "AddServerKeyArg pattern fail to run?";
+  }
+  return *it;
+}
+
+template <class Op>
+struct GenericOpPattern : public OpConversionPattern<Op> {
+  using OpConversionPattern<Op>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      Op op, typename Op::Adaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    SmallVector<Type> retTypes;
+    if (failed(this->getTypeConverter()->convertTypes(op->getResultTypes(),
+                                                      retTypes)))
+      return failure();
+    rewriter.replaceOpWithNewOp<Op>(op, retTypes, adaptor.getOperands(),
+                                    op->getAttrs());
+
+    return success();
+  }
+};
+
+/// Convert a func by adding a server key argument. Converted ops in other
+/// patterns need a server key SSA value available, so this pattern needs a
+/// higher benefit.
+struct AddJaxiteServerKeyArg : public OpConversionPattern<func::FuncOp> {
+  AddJaxiteServerKeyArg(mlir::MLIRContext *context)
+      : OpConversionPattern<func::FuncOp>(context, /* benefit= */ 1000) {}
+
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      func::FuncOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    if (!containsCGGIJaxiteOps(op)) {
+      return failure();
+    }
+
+    auto serverKeyType = jaxite::ServerKeySetType::get(getContext());
+    auto paramsType = jaxite::ParamsType::get(getContext());
+    FunctionType originalType = op.getFunctionType();
+    llvm::SmallVector<Type, 4> newTypes;
+    newTypes.reserve(originalType.getNumInputs() + 2);
+    for (auto t : originalType.getInputs()) {
+      newTypes.push_back(t);
+    }
+    newTypes.push_back(serverKeyType);
+    newTypes.push_back(paramsType);
+    auto newFuncType =
+        FunctionType::get(getContext(), newTypes, originalType.getResults());
+    rewriter.modifyOpInPlace(op, [&] {
+      op.setType(newFuncType);
+      // In addition to updating the type signature, we need to update the
+      // entry block's arguments to match the type signature
+      Block &block = op.getBody().getBlocks().front();
+      block.addArguments(serverKeyType, op.getLoc());
+      block.addArguments(paramsType, op.getLoc());
```

Comment:
addArguments takes a type range so you could just `addArguments({serverKeyType, paramsType}, {loc, loc})`

Reply:
Done